README.md
# ProjectTasks
Package for standardizing plots and tables across projects.


NEWS
NEWS
================
<Erik.Leppo@tetratech.com>

<!-- NEWS.md is generated from NEWS.Rmd. Please edit that file -->

    #> Last Update: 2023-10-06 13:18:37.938269

# ProjectTasks 0.2.0.9003 (2023-10-06)

- break: Added boolean for explicit screening values in legend
  - If didn’t specify parameters and assumed an order will break code

# ProjectTasks 0.2.0.9002 (2023-10-06)

- refactor: CDF plots change SV thickness and color

# ProjectTasks 0.2.0.9001 (2023-10-04)

- break: Added 5th screening value
  - If didn’t specify parameters and assumed an order will break code

# ProjectTasks 0.1.0.9009 (2023-07-25)

- refactor: Add intersect parameter to upset plot function

# ProjectTasks 0.1.0.9008 (2023-07-24)

- refactor: Add pass through parameter (…) to upset plot function

# ProjectTasks 0.1.0.9007 (2023-07-19)

- refactor: Update upset plot function
- feat: Add data for use with upset plot
- refactor: Modify name of map data to better align with upset plot data

# ProjectTasks 0.1.0.9006 (2023-07-19)

- refactor: Update upset plot function

# ProjectTasks 0.1.0.9005 (2023-07-19)

- feat: Add Great Lakes map function
- docs: Update DESCRIPTION for needed packages
- feat: Add example data for Great Lakes map function
- docs: Document example dataset

# ProjectTasks 0.1.0.9004 (2023-04-13)

- feat: Add upset plots (fish tissue)
- docs: Add ComplexUpset and ggplot2 to DESCRIPTION Imports
- style: Update plot_cdf_fishtissue
- fix: Update plot_cdf_fishtissue boo_log_x to fun.xlog

# ProjectTasks 0.1.0.9003 (2022-04-14)

- fix: Update functions with Roxygen header and export
  - msg_status
  - set_xmin_fishtissue
- docs: Add import packages to DESCRIPTION

# ProjectTasks 0.1.0.9002 (2022-04-14)

- feat: Add new functions and helper functions
  - plot CDF (fish tissue)
  - set X min (fish tissue)
  - status message

# ProjectTasks 0.1.0.9001 (2022-04-14)

- Initial commit.
  - NEWS.Rmd
  - README.Rmd
  - DESCRIPTION
  - Library Creation helper notebook

# ProjectTasks 0.1.0 (2022-04-14)

- Repo created on GitHub


R/data.R
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# data_gl2010_map ####
#' @title Great Lakes 2010 PFAS Map Data
#'
#' @description Fish tissue data from Great Lakes 2010 survey.
#' Formatted for use with map function.
#' Analytes include PFAS only.
#' Data set is dated 2021-04-26.
#'
#' @format A data frame with 2041 observations on the following 32 variables.
#' \describe{
#'    \item{\samp{Site ID}}{a character vector}
#'    \item{\samp{EPA Sample ID}}{a character vector}
#'    \item{\samp{EPA Region}}{a character vector}
#'    \item{\code{State}}{a character vector}
#'    \item{\samp{Lake Name}}{a character vector}
#'    \item{\code{Analyte}}{a character vector}
#'    \item{\samp{CAS Number}}{a character vector}
#'    \item{\code{Amount}}{a numeric vector}
#'    \item{\samp{Pre-2018 MN HH 1 Meal/Week 40 ppb Screening Value (BW 70 kg)}}{a numeric vector}
#'    \item{\samp{Exceeds Pre-2018 MN HH 1 Meal/Week 40 ppb SV?}}{a character vector}
#'    \item{\samp{EPA HH 68 ppb Screening Value (75 kg BW; 22 g/day FCR)}}{a numeric vector}
#'    \item{\samp{Exceeds EPA HH 68 ppb SV?}}{a character vector}
#'    \item{\samp{EPA HH 50 ppb Screening Value (75 kg BW; 29.8 g/day FCR)}}{a numeric vector}
#'    \item{\samp{Exceeds EPA HH 50 ppb SV?}}{a character vector}
#'    \item{\samp{EPA HH 46 ppb Screening Value (75 kg BW; 32.4 g/day FCR)}}{a numeric vector}
#'    \item{\samp{Exceeds EPA HH 46 ppb SV?}}{a character vector}
#'    \item{\samp{EPA HH 37 ppb Screening Value (75 kg BW; 40.5 g/day FCR)}}{a numeric vector}
#'    \item{\samp{Exceeds EPA HH 37 ppb SV?}}{a character vector}
#'    \item{\samp{EPA HH 31 ppb Screening Value (75 kg BW; 48.6 g/day FCR)}}{a numeric vector}
#'    \item{\samp{Exceeds EPA HH 31 ppb SV?}}{a character vector}
#'    \item{\samp{EPA HH 8.6 ppb Screening Value (75 kg BW; 175 g/day FCR)}}{a numeric vector}
#'    \item{\samp{Exceeds EPA HH 8.6 ppb SV?}}{a character vector}
#'    \item{\samp{Unit 1}}{a character vector}
#'    \item{\samp{Unit 2}}{a character vector}
#'    \item{\code{MDL}}{a numeric vector}
#'    \item{\code{QL}}{a numeric vector}
#'    \item{\samp{Lab Qualifier Flag}}{a character vector}
#'    \item{\samp{SCC Code}}{a character vector}
#'    \item{\code{Comments}}{a character vector}
#'    \item{\samp{Sort order}}{a numeric vector}
#'    \item{\code{Latitude}}{a numeric vector}
#'    \item{\code{Longitude}}{a numeric vector}
#' }
"data_gl2010_map"
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# data_gl2010_upset ####
#' @title Great Lakes 2010 Multiple Analyte Upset Plot Data
#'
#' @description Fish tissue data from Great Lakes 2010 survey.
#' Formatted for use with update plot function.
#' Analytes include mercury, PFAS, and total PCBs.
#' Exceedance occurrence using screening values (ppb), respectively, 300, 0.52
#' , and 12.
#' Data set is dated 2021-04-26.
#'
#' @format A data frame with 157 observations on the following 4 variables.
#' \describe{
#'    \item{\code{EPA.Sample.ID}}{a character vector}
#'    \item{\code{Mercury}}{a numeric vector}
#'    \item{\code{Perfluorooctanesulfonate..PFOS.}}{a numeric vector}
#'    \item{\code{Total.PCBs}}{a numeric vector}
#' }
"data_gl2010_map"
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


R/msg_status.R
#' @title Status Message
#'
#' @description Print Status Message to Console
#'
#' @details Function for informing user of progress
#'
#' Text is composed using `print` so will have to flush console if used inside
#' of a function or loop.
#'
#' @param fun.status Status of item
#' @param fun.item.num.current Item number
#' @param fun.item.num.total Total number of items
#' @param fun.item.name Name of item
#'
#' @return Returns a text to the console
#'
#' @export
msg_status <- function(fun.status
                           , fun.item.num.current
                           , fun.item.num.total
                           , fun.item.name) {
  print(paste("Processing item "
              , fun.item.num.current
              , " of "
              , fun.item.num.total
              , ", "
              , fun.status
              , ", "
              , fun.item.name
              , "."
              , sep = ""))
} ## FUNCTION.END


R/plot_cdf_fishtissue.R
#' @title Plot Fish Study CDF
#'
#' @description CDF for fish study projects
#'
#' @details Plots CDF with different options for fish tissue studies (NLA,
#' NRSA, and Great Lakes).
#'
#' Plots are base R.
#'
#' Saves plots to a user defined folder.
#'
#' uses the Hmisc and plotrix packages
#'
#' BREAKING CHANGE
#' 2023-10-04
#' Added 5th Screening Value
#' If didn't specify parameters and assumed an order will break code
#'
#' # CDF plotting function for Fish Tissue project
#' # 20161005, Erik.Leppo@tetratech.com
#' # move function to its own script
#' # 20200401, edits
#' ## Screening Value to "Human Health Benchmark", leg_SV_str
#' ## add ng/g to parenthetical, leg_SV_units, adds if present
#' # 20210826, Add legend label update
#' # 20220322, Legend, "Data" to "Fillet Concentration Data"
#' # 20220414, move to package scripts, was fun.plot.CDF.fish
#' # 20231006
#' ## increase SV thickness 1.5 to 2 and color green to darkgreen
#' ## trigger for SV annotations in legend
#'
#' # Derived from NRSA version (more advanced with Log scale than GL version)
#'
#' @param fun.DF Data data frame
#' @param fun.Main plot(main)
#' @param fun.xlim.x plot(xlim), value 1
#' @param fun.xlim.y plot(xlim), value 2
#' @param fun.xlab plot(xlab)
#' @param fun.ylab plot(ylab)
#' @param fun.axis.4.at axis 4 (right), at
#' @param fun.axis.4.labels axis 4 (right), labels
#' @param fun.axis.4.mtext axis 4 (right), margin text label
#' @param fun.SV.plot Screening Value 1, boolean value to plot SV
#' @param fun.SV Screening Value 1, numeric value
#' @param fun.SV.int Screening Value 1, y-intercept value
#' @param fun.SV.lab Screening Value 1, legend (source)
#' @param fun.SV2.plot Screening Value 2, boolean value to plot SV
#' @param fun.SV2 Screening Value 2, numeric value
#' @param fun.SV2.int Screening Value 2, y-intercept value
#' @param fun.SV2.lab Screening Value 2, legend (source)
#' @param fun.SV3.plot Screening Value 3, boolean value to plot SV
#' @param fun.SV3 Screening Value 3, numeric value
#' @param fun.SV3.int Screening Value 3, y-intercept value
#' @param fun.SV3.lab Screening Value 3, legend (source)
#' @param fun.SV4.plot Screening Value 4, boolean value to plot SV
#' @param fun.SV4 Screening Value 4, numeric value
#' @param fun.SV4.int Screening Value 4, y-intercept value
#' @param fun.SV4.lab Screening Value 4, legend (source)
#' @param fun.SV5.plot Screening Value 4, boolean value to plot SV
#' @param fun.SV5 Screening Value 4, numeric value
#' @param fun.SV5.int Screening Value 4, y-intercept value
#' @param fun.SV5.lab Screening Value 4, legend (source)
#' @param fun.break.boo Should x-axis include a break. Default = FALSE
#' @param fun.break.max.at . Default = NA
#' @param fun.break.pos location of break. Default = NA
#' @param fun.break.axis.at break. Default = NA
#' @param fun.break.axis.labels new x-axis labels with break
#' @param leg_SV_str Legend screening value, SV text.  Default = "Screening Value"
#' @param leg_SV_units Legend screening value, units.  Default = NA
#' @param leg_SV_verbose Boolean to include SV in the legend with or without
#' lab, units, and SV value.  Default = TRUE (FALSE will be leg_SV_str only)
#' @param fun.axis.2.mtext axis 2 (left) margin text label. Default = NA
#' @param fun.xlog Boolean value if x-axis should be log10 transformed.  Default = FALSE
#'
#' @return CDF plot customized for Fish Tissue project.
#'
#' @examples
#' # non at this time
#'
#' @export
plot_cdf_fishtissue <- function(fun.DF
                                , fun.Main
                                , fun.xlim.x
                                , fun.xlim.y
                                , fun.xlab
                                , fun.ylab
                                , fun.axis.4.at
                                , fun.axis.4.labels
                                , fun.axis.4.mtext
                                , fun.SV.plot
                                , fun.SV
                                , fun.SV.int
                                , fun.SV.lab
                                , fun.SV2.plot
                                , fun.SV2
                                , fun.SV2.int
                                , fun.SV2.lab
                                , fun.SV3.plot
                                , fun.SV3
                                , fun.SV3.int
                                , fun.SV3.lab
                                , fun.SV4.plot
                                , fun.SV4
                                , fun.SV4.int
                                , fun.SV4.lab
                                , fun.SV5.plot
                                , fun.SV5
                                , fun.SV5.int
                                , fun.SV5.lab
                                , fun.break.boo = FALSE
                                , fun.break.max.at = NA
                                , fun.break.max.text = NA
                                , fun.break.pos = NA
                                , fun.break.axis.at = NA
                                , fun.break.axis.labels = NA
                                , leg_SV_str = "Screening Value"
                                , leg_SV_units = NA
                                , leg_SV_verbose = TRUE
                                , fun.axis.2.mtext = NA
                                , fun.xlog = FALSE
                                ) {
  #browser()
  # define some values ----
  lwd.data <- 2.5
  lwd.SV <- 2 # 1.5
  lwd.CI <- 1.5 #lwd.SV
  col.data <- "blue"
  col.SV <- "darkgreen" # "green"
  col.CI <- "dark red"
  lty.data <- "solid"
  lty.SV <- "dashed"
  lty.CI <- "dotted"
  mtext.line <- 3 # 20210827, added for axis 2 and 4 mtext
  mtext.cex <- 1.2
  # create a graph with Value, ensure values are sorted; format graph and send to file
  plot.val.max <- max(fun.DF$Value)
  #plot.ylim.max <- 101
  #plot.xlim.min <-   # use formula based on max(fun.myDF$Value)
  par(oma = c(0, 0, 0, 3))
  par(xaxs = "i", yaxs = "i", cex.main = 1.4, cex.lab = 1.2)
  #

  #  browser()
  # log10_x, mod data ----
  if (fun.xlog == TRUE) {
    # Munge data for plot
    #fun.DF$Value <- log10(fun.DF$Value)
    # SV segments values
    # SV.val.seg <- log10(fun.SV)
    # SV2.val.seg <- log10(fun.SV2)
    # SV3.val.seg <- log10(fun.SV3)
    # SV4.val.seg <- log10(fun.SV4)
    seg.h.x0 <- 0.001
    seg.v.y0 <- -10
  } else {
    # SV.val.seg <- fun.SV
    # SV2.val.seg <- fun.SV2
    # SV3.val.seg <- fun.SV3
    # SV4.val.seg <- fun.SV4
    seg.h.x0 <- 0
    seg.v.y0 <- 0
  }## IF ~ boo_log10_x

  # Segment values
  SV.val.seg <- fun.SV
  SV2.val.seg <- fun.SV2
  SV3.val.seg <- fun.SV3
  SV4.val.seg <- fun.SV4
  SV5.val.seg <- fun.SV5

  # Legend SV value
  SV.val.leg <- fun.SV
  SV2.val.leg <- fun.SV2
  SV3.val.leg <- fun.SV3
  SV4.val.seg <- fun.SV4
  SV5.val.seg <- fun.SV5


  # Plot ----
  # without axes so can manipulate for BREAK
  if (fun.break.boo == TRUE) {##IF.fun.break.boo.START
    ## plot, break, TRUE ----
    plot(fun.DF$Estimate.P~sort(fun.DF$Value)
         , ylim = c(0,101)
         , xlim = c(fun.xlim.x, 1.01 * (fun.break.max.at))
         , type = "l"
         , col = col.data
         , lwd = lwd.data
         , lty = lty.data
         , main = fun.Main
         , xlab = fun.xlab
         , ylab = fun.ylab
         , axes = FALSE)
    #
    grid(col = "lightgray", lty = "dotted", lwd = 0.5)
    box()
    axis(1, at = fun.break.axis.at, label = fun.break.axis.labels)
    plotrix::axis.break(1, fun.break.pos, style = "zigzag")
  } else {
    ## plot, break, FALSE ----
    if (fun.xlog == TRUE) {
      ## plot, log10----
      plot(fun.DF$Estimate.P~sort(fun.DF$Value)
           , ylim = c(0,101)
           #, xlim = c(fun.xlim.x, 1.01 * (fun.xlim.y))
           , type = "l"
           , col = col.data
           , lwd = lwd.data
           , lty = lty.data
           , main = fun.Main
           , xlab = fun.xlab
           , ylab = fun.ylab
           , axes = FALSE
           , log = "x")
      ### Axis 1 (bottom) ----
      # idea
      # https://stackoverflow.com/questions/6897243/labelling-logarithmic-scale-display-in-r
      axis_log10_at <- outer(c((1:9)/100, (1:9)/10, 1:9), 10^(0:4))
      axis_log10_lab <- ifelse(log10(axis_log10_at) %% 1 == 0, axis_log10_at, NA)
      # Ticks, Minor
      axis(1, at = axis_log10_at, labels = axis_log10_lab, tck = -0.01)
      # Ticks, Major
      axis(1, at = axis_log10_lab, labels = NA, tck = -0.025)
    } else {
      # plot, base ----
      plot(fun.DF$Estimate.P~sort(fun.DF$Value)
           , ylim = c(0,101)
           , xlim = c(fun.xlim.x,1.01 * (fun.xlim.y))
           , type = "l"
           , col = col.data
           , lwd = lwd.data
           , lty = lty.data
           , main = fun.Main
           , xlab = fun.xlab
           , ylab = fun.ylab
           , axes = FALSE)
      ### Axis 1 (bottom) ----
      axis(1)
    }## IF ~ boo_log10_x ~ 2
    grid(col = "lightgray", lty = "dotted", lwd = 0.5)
    box()
  }##IF.fun.break.boo.START

  # # Log X
  # plot(fun.DF$Estimate.P~sort(fun.DF$Value)
  #      , ylim=c(0,101), log = "x"
  #      , type="l", col=col.data, lwd=lwd.data, lty=lty.data
  #      , main=fun.Main, xlab=fun.xlab, ylab=fun.ylab, axes=FALSE)
  # grid(col="lightgray",lty="dotted",lwd=0.5)
  # xlog_ticks_at <- axTicks(1)
  # # xlog_ticks_lab <- sapply(xlog_ticks_at, function(i)
  # #                         as.expression(bquote(10^ .(i)))
  # #                         )
  # xlog_ticks_lab <- format(xlog_ticks_at, scientific = FALSE, trim = TRUE)
  # axis(1, at = xlog_ticks_at, labels = xlog_ticks_lab)
  #
  ## used log="x" inside plot() so above not needed.

  box() # covers up grid lines

  # (moved grid and box inside IF/THEN)

  # Axis 2 (left) ----
  axis(2)
  # Add minor tickmarks (1 = no tick, ratio = compared to major ticks)
  Hmisc::minor.tick(nx = 1, ny = 4, tick.ratio = 0.33)

  # Axis 4 (right) ----
  axis(4, at = fun.axis.4.at, labels = fun.axis.4.labels)

  # 95% confidence intervals ----
  lines(fun.DF$UCB95Pct.P ~ sort(fun.DF$Value)
        , col = col.CI
        , lty = lty.CI
        , lwd = lwd.CI)
  lines(fun.DF$LCB95Pct.P ~ sort(fun.DF$Value)
        , col = col.CI
        , lty = lty.CI
        , lwd = lwd.CI)

  # Screening Value Lines (vertical & horizontal) ----
  mySV.num <- 0
  #seg.x0 defined under log10_x
  # SV1 - 20110713, add back in, only plot if "fun.SV.plot=TRUE".
  if (fun.SV.plot == TRUE) { ## IF.START
    mySV.num <- 1
    segments(SV.val.seg
             , seg.v.y0
             , SV.val.seg
             , fun.SV.int
             , col = col.SV
             , lwd = lwd.SV
             , lty = lty.SV)
    segments(seg.h.x0
             , fun.SV.int
             , SV.val.seg
             , fun.SV.int
             , col = col.SV
             , lwd = lwd.SV
             , lty = lty.SV)
  } ## IF.END
  # SV2 - 20161004, extra SV plot
  if (fun.SV2.plot == TRUE) { ## IF.START
    mySV.num <- 2
    segments(SV2.val.seg
             , seg.v.y0
             , SV2.val.seg
             , fun.SV2.int
             , col = col.SV
             , lwd = lwd.SV
             , lty = lty.SV)
    segments(seg.h.x0
             , fun.SV2.int
             , SV2.val.seg
             , fun.SV2.int
             , col = col.SV
             , lwd = lwd.SV
             , lty = lty.SV)
  } ## IF.END
  # SV3 - 20161006, extra SV plot
  if (fun.SV3.plot == TRUE) { ## IF.START
    mySV.num <- 3
    segments(SV3.val.seg
             , seg.v.y0
             , SV3.val.seg
             , fun.SV3.int
             , col = col.SV
             , lwd = lwd.SV
             , lty = lty.SV)
    segments(seg.h.x0
             , fun.SV3.int
             , SV3.val.seg
             , fun.SV3.int
             , col = col.SV
             , lwd = lwd.SV
             , lty = lty.SV)
  } ## IF.END
  # SV4 - 20161006, extra SV plot
  if (fun.SV4.plot == TRUE) { ## IF.START
    mySV.num <- 4
    segments(SV4.val.seg
             , seg.v.y0
             , SV4.val.seg
             , fun.SV4.int
             , col = col.SV
             , lwd = lwd.SV
             , lty = lty.SV)
    segments(seg.h.x0
             , fun.SV4.int
             , SV4.val.seg
             , fun.SV4.int
             , col = col.SV
             , lwd = lwd.SV
             , lty = lty.SV)
  } ## IF.END
  # SV5 - 20231004, extra SV plot
  if (fun.SV5.plot == TRUE) {
    mySV.num <- 5
    segments(SV5.val.seg
             , seg.v.y0
             , SV5.val.seg
             , fun.SV5.int
             , col = col.SV
             , lwd = lwd.SV
             , lty = lty.SV)
    segments(seg.h.x0
             , fun.SV5.int
             , SV5.val.seg
             , fun.SV5.int
             , col = col.SV
             , lwd = lwd.SV
             , lty = lty.SV)
  } ## IF.END
  #
  # Legend ----
  # (EWL, 20120830, added SV legend back (inside of SV IF/THEN)
  # highest number will take priority
  # 20200401, add variable for legend text and units

  if (!is.na(leg_SV_units)) {
    fun.SV  <- paste0(fun.SV,  " ", leg_SV_units)
    fun.SV2 <- paste0(fun.SV2, " ", leg_SV_units)
    fun.SV3 <- paste0(fun.SV3, " ", leg_SV_units)
    fun.SV4 <- paste0(fun.SV4, " ", leg_SV_units)
    fun.SV5 <- paste0(fun.SV5, " ", leg_SV_units)
  }## IF ~ is.na(leg_SV_units) ~ END

  # Modify legend label
  fun.SV.lab  <- ifelse(is.na(fun.SV.lab), "", paste0(fun.SV.lab, ", "))
  fun.SV2.lab <- ifelse(is.na(fun.SV2.lab), "", paste0(fun.SV2.lab, ", "))
  fun.SV3.lab <- ifelse(is.na(fun.SV3.lab), "", paste0(fun.SV3.lab, ", "))
  fun.SV4.lab <- ifelse(is.na(fun.SV4.lab), "", paste0(fun.SV4.lab, ", "))
  fun.SV5.lab <- ifelse(is.na(fun.SV5.lab), "", paste0(fun.SV5.lab, ", "))

  if (leg_SV_verbose == FALSE) {
    legend("bottomright"
           , lty = c(lty.data, lty.SV, lty.CI)
           , lwd = c(lwd.data, lwd.SV, lwd.CI)
           , col = c(col.data, col.SV, col.CI)
           , legend = c("Fillet Concentration Data"
                        , leg_SV_str
                        , "95% Confidence Intervals")
           , bg = "white")
  } else if (mySV.num == 1) { ##IF.START
    legend("bottomright"
           , lty = c(lty.data, lty.SV, lty.CI)
           , lwd = c(lwd.data, lwd.SV, lwd.CI)
           , col = c(col.data, col.SV, col.CI)
           , legend = c("Fillet Concentration Data"
                     , paste(leg_SV_str, " (", fun.SV.lab, fun.SV, ")", sep = "")
                     , "95% Confidence Intervals")
           , bg = "white")
  } else if (mySV.num == 2) {
    legend("bottomright"
           , lty = c(lty.data, rep(lty.SV, mySV.num), lty.CI)
           , lwd = c(lwd.data, rep(lwd.SV, mySV.num), lwd.CI)
           , col = c(col.data, rep(col.SV, mySV.num), col.CI)
           , legend = c("Fillet Concentration Data"
                     , paste(leg_SV_str, " (", fun.SV.lab, fun.SV, ")", sep = "")
                     , paste(leg_SV_str, " (", fun.SV2.lab, fun.SV2, ")", sep = "")
                     , "95% Confidence Intervals")
           , bg = "white")

  } else if (mySV.num == 3) {
    legend("bottomright"
           , lty = c(lty.data,rep(lty.SV, mySV.num), lty.CI)
           , lwd = c(lwd.data,rep(lwd.SV, mySV.num), lwd.CI)
           , col = c(col.data,rep(col.SV, mySV.num), col.CI)
           , legend = c("Fillet Concentration Data"
                     , paste(leg_SV_str, " (", fun.SV.lab, fun.SV, ")", sep = "")
                     , paste(leg_SV_str, " (", fun.SV2.lab, fun.SV2, ")", sep = "")
                     , paste(leg_SV_str, " (", fun.SV3.lab, fun.SV3, ")", sep = "")
                     , "95% Confidence Intervals")
           ,bg = "white")
  } else if (mySV.num == 4) {
    legend("bottomright"
           , lty = c(lty.data,rep(lty.SV, mySV.num), lty.CI)
           , lwd = c(lwd.data,rep(lwd.SV, mySV.num), lwd.CI)
           , col = c(col.data,rep(col.SV, mySV.num), col.CI)
           , legend = c("Fillet Concentration Data"
                     , paste(leg_SV_str, " (", fun.SV.lab, fun.SV, ")", sep = "")
                     , paste(leg_SV_str, " (", fun.SV2.lab, fun.SV2, ")", sep = "")
                     , paste(leg_SV_str, " (", fun.SV3.lab, fun.SV3, ")", sep = "")
                     , paste(leg_SV_str, " (", fun.SV4.lab, fun.SV4, ")", sep = "")
                     , "95% Confidence Intervals")
           , bg = "white")
  } else if (mySV.num == 5) {
    legend("bottomright"
           , lty = c(lty.data,rep(lty.SV, mySV.num), lty.CI)
           , lwd = c(lwd.data,rep(lwd.SV, mySV.num), lwd.CI)
           , col = c(col.data,rep(col.SV, mySV.num), col.CI)
           , legend = c("Fillet Concentration Data"
                        , paste(leg_SV_str, " (", fun.SV.lab, fun.SV, ")", sep = "")
                        , paste(leg_SV_str, " (", fun.SV2.lab, fun.SV2, ")", sep = "")
                        , paste(leg_SV_str, " (", fun.SV3.lab, fun.SV3, ")", sep = "")
                        , paste(leg_SV_str, " (", fun.SV4.lab, fun.SV4, ")", sep = "")
                        , paste(leg_SV_str, " (", fun.SV5.lab, fun.SV5, ")", sep = "")
                        , "95% Confidence Intervals")
           , bg = "white")
  } else {
    legend("bottomright"
           , lty = c(lty.data,lty.CI)
           , lwd = c(lwd.data,lwd.CI)
           , col = c(col.data,col.CI)
           , legend = c("Fillet Concentration Data"
                     , "95% Confidence Intervals")
           , bg = "white")
  }

  #~~~~~~~~~~~~~~~~~~~~~~~~
  # Break Symbol ----
  # Break (after legend so goes on top)
  if (fun.break.boo == TRUE) {##IF.fun.break.boo.START
    plotrix::axis.break(1, fun.break.pos, style = "zigzag")
  }

  # Axis 2 and 4 Titles ----
  par(new = TRUE)
  # different title based on projet (20161004)
  mtext(fun.axis.4.mtext, side = 4, line = mtext.line, cex = mtext.cex)
  # different yaxis (20210827), when use superscript have to use mtext instead
  if (fun.ylab == "") {
    mtext(fun.axis.2.mtext, side = 2, line = mtext.line - 0.5, cex = mtext.cex)
  }## IF ~ !is.na(fun.axis.2.mtext) ~ END
  #
  # Turn off new ----
  par(new = FALSE)
} ## FUNCTION ~ END


R/plot_upset_fishtissue.R
#' @title Plot Fish Study Upset
#'
#' @description Upset plot for fish study projects
#'
#' @details Upset plot using the `ComplexUpset` package.
#'
#' Upset plots show co-occurrence of different variables.  The left side sizes
#' plot is turned on by default.  This creates extra area in the upper left
#' that can be used for an alternative location for the plot title.  The
#' function is set up to use the basic title as default but has input parameters
#' to change the location (e.g., upper left).
#'
#' A ggplot object is returned that the user can further modify and save.
#' #'
#' # 2023-07-20, moved code to package scripts.
#' Derived for Great Lakes studies.
#'
#' @param data Data as data frame (variables are 1 or 0 as columns).
#' @param intersect Columns to be used to compose the intersection.
#' Default = names(data)[-1]
#' @param title_str Title string. Default = NA
#' @param title_custloc Use a custom location for the title? Default = FALSE
#' @param title_x Title x placement.  Default = 0.2
#' @param title_y Title y placement. Default = 0.9
#' @param title_fontsize Title font size.  Default = 16
#' @param cap_str Caption string. Default = NA
#' @param cap_custloc Use a custom location for the caption? Default = FALSE
#' @param cap_x Caption x placement. Default = 0.5
#' @param cap_y Caption y placement. Default = 0.025
#' @param cap_fontsize Caption font size. Default = 8
#' @param ... Arguments to be passed on to ComplexUpset::upset
#'
#' @return A ggplot object is returned.
#'
#' @examples
#' # Data
#' df_data <- data_gl2010_upset
#'
#' # Parameters ----
#' namedata <- "Great Lakes, 2010"
#' fishconsumer <- "general"
#' sv_hg <- 300
#' sv_pfas <- 0.52
#' sv_pcb <- 12
#'
#' # Plot Parts ----
#' title_str <- paste0(namedata
#'                     , "\n(n = ", nrow(df_data), ")"
#'                     , "\nExceeds Screening Value")
#' cap_str <- paste0("Screening Values for "
#'                   , fishconsumer
#'                   , " fish consumer; "
#'                   , "Mercury = "
#'                   , sv_hg
#'                   , ", "
#'                   , "PFAS = "
#'                   , sv_pfas
#'                   , ", "
#'                   , "Total PCBs = "
#'                   , sv_pcb
#'                   )
#'
#' # Plot, title and caption, base location ----
#' p1 <- plot_upset_fishtissue(df_data = df_data
#'                            , title_str = title_str
#'                            , cap_str = cap_str)
#' p1
#'
#' # Plot, title and caption, custom location ----
#' p2 <- plot_upset_fishtissue(df_data = df_data
#'                            , title_str = title_str
#'                            , title_custloc = TRUE
#'                            , title_x = 0.16
#'                            , title_y = 0.90
#'                            , title_fontsize = 16
#'                            , cap_str = cap_str
#'                            , cap_custloc = TRUE
#'                            , cap_x = 0.5
#'                            , cap_y = 0.025
#'                            , cap_fontsize = 8)
#' p2
#'
#' @export
plot_upset_fishtissue <- function(df_data
                                  , intersect = names(df_data)[-1]
                                  , title_str = NA
                                  , title_custloc = FALSE
                                  , title_x = 0.2
                                  , title_y = 0.9
                                  , title_fontsize = 16
                                  , cap_str = NA
                                  , cap_custloc = FALSE
                                  , cap_x = 0.5
                                  , cap_y = 0.025
                                  , cap_fontsize = 8
                                  , ...) {

  # UpSetR
  # https://upset.app/

  #https://krassowski.github.io/complex-upset/articles/Examples_R.html
  # Complex Upset adds ggplot to UpSetR

  plot_type <- "ComplexUpset"
  plot_type <- tolower(plot_type)

  if (plot_type == "upsetr") {
    # ## UpSetR ----
    # ### Plot ----
    # UpSetR::upset(df_data
    #               , text.scale = c(1.3, 1.3, 1.1, 1.1, 1.2, 1.2)
    #               , line.size = 1.1
    #               , point.size = 2.1
    #               # , mainbar.y.label = "Number of Intersecting Samples"
    #               , set_size.show = TRUE
    #               , set_size.scale_max = nrow(df_data) * 1.1
    #               , order.by = "freq"
    #               , nsets = ncol(df_data) - 1)
    #
    # # text.scale:
    # # c(intersection size title
    # #  , intersection size tick labels
    # #  , set size title
    # #  , set size tick labels
    # #  , set names
    # #  , numbers above bars)
    #
    # ### Title ----
    # grid::grid.text(title_str
    #                 , x = title_x
    #                 , y = title_y
    #                 , gp = grid::gpar(fontsize = title_fontsize))
    #
    # ### Caption ----
    # grid::grid.text(cap_str
    #                 , x = 0.5
    #                 , y = 0.025
    #                 , gp = grid::gpar(fontsize = cap_fontsize))

  } else if (plot_type == "complexupset") {
    ## ComplexUpset----
    ### Plot ----
    p <- ComplexUpset::upset(df_data
                        , intersect = intersect
                        , set_sizes = (ComplexUpset::upset_set_size() +
                              ggplot2::geom_text(ggplot2::aes(label =
                                                    ggplot2::after_stat(count))
                                        , hjust = 1.1
                                        , stat = 'count')
                              + ggplot2::expand_limits(y = nrow(df_data) * 1.15)
                                      )
                        , ...)
    #### Title ----
    if (title_custloc == TRUE) {
      p <- p +
        grid::grid.text(title_str
                        , x = title_x
                        , y = title_y
                        , gp = grid::gpar(fontsize = title_fontsize))
    } else {
      p <- p +
        ggplot2::labs(title = title_str)
    }## IF ~ title_custloc

    #### Caption ----
    if (cap_custloc == TRUE) {
      p <- p +
        grid::grid.text(cap_str
                        , x = cap_x
                        , y = cap_y
                        , gp = grid::gpar(fontsize = cap_fontsize))
    } else {
      p <- p +
        ggplot2::labs(caption = cap_str)
    }## IF ~ title_custloc

    # NOTES ----
    # to get n on set sizes plot
    # https://github.com/krassowski/complex-upset/issues/24
    #
    # https://krassowski.github.io/complex-upset/articles/Examples_R.html
    # about 50% down

  }## IF ~ plot_type


  # Return ----
  print(p)

} ## FUNCTION ~ END
